### PR TITLE
add docments to 33_text.models.core notebook

### LINKED ISSUE #23: 33_text.models.core: create docments
Add documents to [33_text.models.core](https://github.com/muellerzr/fastai-docment-sprint/blob/master/nbs/33_text.models.core.ipynb).

See the style guide and contribution guide for more details (links will be added when created).

This requires a moderate familiarity with pytorch needed. This notebook defines how to create the core components for language and text classification models 

Documents progress(only one section):

- [ ] Language models
- [ ] Classification models

### PR DESCRIPTION
**Pull request by arampacha**
closes #23

### CODE REVIEW

File: fastai/text/models/core.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from ...data.all import *
 from ..core import *
 from .awdlstm import *
+from typing import List
```
Comment (muellerzr) [inline fastai/text/models/core.py:10]:
```suggestion
```
We don't use typing

File: fastai/text/models/core.py [hunk 2]
```diff
@@ -42,7 +49,12 @@ def reset(self):
         for c in self.children(): getattr(c, 'reset', noop)()
 
 # Cell
-def get_language_model(arch, vocab_sz, config=None, drop_mult=1.):
+def get_language_model(
+    arch:type, # Class defining language model architecture
```
Comment (muellerzr) [inline fastai/text/models/core.py:53]:
```suggestion
    arch:callable, # Function or class that can generate a language model architecture
```

File: fastai/text/models/core.py [hunk 3]
```diff
@@ -57,14 +69,19 @@ def get_language_model(arch, vocab_sz, config=None, drop_mult=1.):
...
 # Cell
 class SentenceEncoder(Module):
     "Create an encoder over `module` that can process a full sentence."
-    def __init__(self, bptt, module, pad_idx=1, max_len=None): store_attr('bptt,module,pad_idx,max_len')
+    def __init__(self,
+        bptt:int, # Backpropagation through time
+        module:nn.Module, # Wrapped module
+        pad_idx:int=1, # Padding token id
+        max_len:int=None # Maximal output length
+    ): store_attr('bptt,module,pad_idx,max_len')
```
Comment (muellerzr) [inline fastai/text/models/core.py:84]:
```suggestion
    ): 
        store_attr('bptt,module,pad_idx,max_len')
```

File: fastai/text/models/core.py [hunk 4]
```diff
@@ -84,7 +101,11 @@ def forward(self, input):
         return outs,mask
 
 # Cell
-def masked_concat_pool(output, mask, bptt):
+def masked_concat_pool(
+    output:Tensor, # Output of sentence encoder
+    mask:Tensor, # Boolian mask as returned by sentence encoder
```
Comment (muellerzr) [inline fastai/text/models/core.py:106]:
```suggestion
    mask:Tensor, # Boolean mask as returned by sentence encoder
```

File: fastai/text/models/core.py [hunk 5]
```diff
@@ -84,7 +101,11 @@ def forward(self, input):
...
+    bptt:int # Backpropagation through time
+) -> Tensor: # Concattenation of [last_hidden, max_pool, avg_pool]
```
Comment (muellerzr) [inline fastai/text/models/core.py:108]:
```suggestion
) -> Tensor: # Concatenation of [last_hidden, max_pool, avg_pool]
```

File: fastai/text/models/core.py [hunk 6]
```diff
@@ -97,7 +118,12 @@ def masked_concat_pool(output, mask, bptt):
 # Cell
 class PoolingLinearClassifier(Module):
     "Create a linear classifier with pooling"
-    def __init__(self, dims, ps, bptt, y_range=None):
+    def __init__(self,
+        dims:List[int], # List of hidden sizes for MLP
+        ps:List[float], # List of dropout probabilities
```
Comment (muellerzr) [inline fastai/text/models/core.py:123]:
```suggestion
        dims:list, # List of hidden sizes for MLP as `int`s
        ps:list, # List of dropout probabilities as `float`s
```

File: fastai/text/models/core.py [hunk 7]
```diff
@@ -112,8 +138,19 @@ def forward(self, input):
         return x, out, out
 
 # Cell
-def get_text_classifier(arch, vocab_sz, n_class, seq_len=72, config=None, drop_mult=1., lin_ftrs=None,
-                        ps=None, pad_idx=1, max_len=72*20, y_range=None):
+def get_text_classifier(
+    arch:type, # Class defining language model architecture
```
Comment (muellerzr) [inline fastai/text/models/core.py:142]: Same nit as earlier w/ `callable` type + string

File: fastai/text/models/core.py [hunk 8]
```diff
@@ -112,8 +138,19 @@ def forward(self, input):
...
+    vocab_sz:int, # Size of the vocabulary
+    n_class:int, # Number of classes
+    seq_len:int=72, # Backpropagation through time
+    config:dict=None, # Encoder configuration dictionary
+    drop_mult:float=1., # Multiplicative factor to scale all dropout probabilities in `config`
+    lin_ftrs:List[int]=None, # List of hidden sizes for classifier head
+    ps:List[float]=None, # List of dropout probabilities for classifier head
```
Comment (muellerzr) [inline fastai/text/models/core.py:149]: Remove typing and format similar to previous suggestion

**Review by muellerzr — CHANGES_REQUESTED** [review]
Left some nits, great start!

**Review by marii-moe — CHANGES_REQUESTED** [review]
Joined this one because another Reviewer was needed. Only one requested change

**Review by marii-moe — APPROVED** [review]
LGTM

File: fastai/text/models/core.py [hunk 9]
```diff
@@ -42,7 +48,12 @@ def reset(self):
         for c in self.children(): getattr(c, 'reset', noop)()
 
 # Cell
-def get_language_model(arch, vocab_sz, config=None, drop_mult=1.):
+def get_language_model(
+    arch:callable, # Function or class that can generate a language model architecture
```
Comment (muellerzr) [inline fastai/text/models/core.py:52]:
```suggestion
    arch, # Function or class that can generate a language model architecture
```
After talking with Jeremy, let's get rid of this type hint. The docment says enough

File: fastai/text/models/core.py [hunk 10]
```diff
@@ -112,8 +139,19 @@ def forward(self, input):
         return x, out, out
 
 # Cell
-def get_text_classifier(arch, vocab_sz, n_class, seq_len=72, config=None, drop_mult=1., lin_ftrs=None,
-                        ps=None, pad_idx=1, max_len=72*20, y_range=None):
+def get_text_classifier(
+    arch:callable, # Function or class that can generate a language model architecture
```
Comment (muellerzr) [inline fastai/text/models/core.py:143]:
```suggestion
    arch, # Function or class that can generate a language model architecture
```

**Review by muellerzr — APPROVED** [review]
One small change due to new information and we're good!

**Comment by arampacha** [comment]
sorry for a bit of a mess, I've got a lot of distraction lately

**Pull request merged**